### PR TITLE
OdbcType implementation for u8 vector

### PR DESCRIPTION
**Pull request by V-0-1-D**
OdbcType implementation for u8 vector was added in order to receive non utf8 strings.

### CODE REVIEW

**Comment by Koka** [comment]
Thanks for your contribution!
Please consider adding corresponding test to `/tests/native_types.rs`.

We have some plans on handling non utf8 strings in a different way, but I think we could add this PR as a temporary solution (as it was suggested in #52)

File: src/statement/types.rs [hunk 1]
```diff
@@ -202,3 +202,22 @@ unsafe impl<'a> OdbcType<'a> for f64 {
...
+    fn c_data_type() -> ffi::SqlCDataType {
+        ffi::SQL_C_CHAR
+    }
+
+    fn convert(buffer: &'a [u8]) -> Self {
+        buffer.to_vec()
+    }
+
+    fn column_size(&self) -> ffi::SQLULEN {
+        self.len() as ffi::SQLULEN
```
Comment (pacman82) [inline src/statement/types.rs]: This should be something like:
```rust
std::cmp::min(self.len(), SQLULEN::max_value() as usize) as SQLULEN
```

File: src/statement/types.rs [hunk 2]
```diff
@@ -202,3 +202,22 @@ unsafe impl<'a> OdbcType<'a> for f64 {
     fn column_size(&self) -> ffi::SQLULEN { size_of::<Self>() as ffi::SQLULEN }
     fn value_ptr(&self) -> ffi::SQLPOINTER { self as *const Self as ffi::SQLPOINTER }
 }
+
+unsafe impl<'a> OdbcType<'a> for Vec<u8> {
+    fn sql_data_type() -> ffi::SqlDataType { ffi::SQL_VARCHAR }
```
Comment (pacman82) [inline src/statement/types.rs]: This should rather be `SQL_VARBINARY`, so no encoding is assumed. It's a design error to have the Data Type being part of tied trait anyway, but for now, I would just change this to `SQL_VARBINARY` to prevent the driver and driver manager from performing transformations on the data send through `Vec<u8>`.

File: src/statement/types.rs [hunk 3]
```diff
@@ -202,3 +202,22 @@ unsafe impl<'a> OdbcType<'a> for f64 {
...
+    fn convert(buffer: &'a [u8]) -> Self {
+        buffer.to_vec()
+    }
+
+    fn column_size(&self) -> ffi::SQLULEN {
+        self.len() as ffi::SQLULEN
+    }
+
+    fn value_ptr(&self) -> ffi::SQLPOINTER {
+        self.as_ptr() as *const Self as ffi::SQLPOINTER
```
Comment (pacman82) [inline src/statement/types.rs]: This line will likely break for empty `Vec`. Return `null_mut()` for these

File: src/statement/types.rs [hunk 4]
```diff
@@ -202,3 +202,22 @@ unsafe impl<'a> OdbcType<'a> for f64 {
     fn column_size(&self) -> ffi::SQLULEN { size_of::<Self>() as ffi::SQLULEN }
     fn value_ptr(&self) -> ffi::SQLPOINTER { self as *const Self as ffi::SQLPOINTER }
 }
+
+unsafe impl<'a> OdbcType<'a> for Vec<u8> {
+    fn sql_data_type() -> ffi::SqlDataType { ffi::SQL_VARCHAR }
+    fn c_data_type() -> ffi::SqlCDataType {
+        ffi::SQL_C_CHAR
```
Comment (pacman82) [inline src/statement/types.rs]: SQL_C_BINARY would be preferable. Vectors can contain `0` after all.

File: tests/native_types.rs [hunk 5]
```diff
@@ -86,3 +86,8 @@ fn _f32() {
 fn _f64() {
     test_type!(f64, C, 3.14)
 }
+
+#[test]
+fn _u8vector() {
+    test_type!(Vec<u8>, A, String::from("Hello, World!").into_bytes())
+}
```
Comment (pacman82) [inline tests/native_types.rs]: Good, but an additional test for an empty `Vec` would really be nice

**Comment by V-0-1-D** [comment]
@pacman82 I've changed the `column_size` function, add the `null_mut()` return and add the test for the empty vector and change types to the `SQL_EXT_VARBINARY` and the `SQL_C_BINARY`, but have an issue with them.
Test of input parameters works with the `SQL_VARCHAR` and the `SQL_C_CHAR` but not works with the `SQL_EXT_VARBINARY` and the `SQL_C_BINARY`. Is this PR is ok, or i should include the input test and use char types? Or maybe i miss something else?

**Comment by pacman82** [comment]
@V-0-1-D Well binary data works a bit different from character data. However (re)reading your description of the PR:

> OdbcType implementation for u8 vector was added in order to receive non utf8 strings.

If you want to receive non-utf8 Strings and not arbitrary binary data, maybe you to support `CString` instead of `Vec<u8>`. `CString` wraps a `Vec<u8>` but actually represents any `CString` (utf8 or not). This expresses the intend clearer and we can still use `Vec<u8>` to receive arbitrary binary data. If you would change your PR from `Vec<u8>` to `CString` I would not object to having the Data Type set to `SQL_VARCHAR` and C Type set to `SQL_C_CHAR`.

Sorry for the inconvenience, I should have suggested this in the first place. Somewhere mid review, the fact that you wanted to receive Text data slipped my mind.

**Comment by Koka** [comment]
Closing this in favor of https://github.com/Koka/odbc-rs/pull/68
